legate/core/communicator.py

```diff
@@ -147,6 +149,12 @@ def _initialize(self, volume: int) -> FutureMap:
             side_effect=True,
         )
         cpucoll_id = task.execute_single()
+        cpucoll_id.wait()
+        coll_uid = native_np.frombuffer(
+            cpucoll_id.get_buffer(), dtype=native_np.int32
+        )[0]
+        self._runtime.issue_execution_fence(block=True)
+        self._runtime.core_library.legate_cpucoll_initcomm(coll_uid)
```

Comment:
hoist this to a task, then you don't need to remove all the previous blocking calls.

Reply:
I removed the `issue_execution_fence`, but kept the `wait`. It is difficult to convert this function into a task because the `MPI_Comm_dup` inside `legate_cpucoll_initcomm` needs to be called by the main thread of each node. Actually the `MPI_Comm_dup` acts like a `MPI_Barrier`, so it will block the calling thread of all nodes, so I do not think it is necessary to convert it into an index task. As long as I remove the `issue_execution_fence`, it will not block the Legion runtime to progress tasks on other threads.

Reply:
> It is difficult to convert this function into a task because the MPI_Comm_dup inside legate_cpucoll_initcomm needs to be called by the main thread of each node.

I don't think this is 100% accurate. I believe what you really want is to call that function only once per node. note that there's no such thing as "main thread" in the legate model. even the python task you're calling that function in isn't really the main thread, but just one of the threads managed by the runtime.

if you want to launch something per node instead of per processor, you just need to adjust the launch domain and update the mapper to do the right thing.

> Actually the MPI_Comm_dup acts like a MPI_Barrier, so it will block the calling thread of all nodes, so I do not think it is necessary to convert it into an index task.

there's a big difference between blocking in one of the leaf tasks and doing it in the main python task; the former only blocks execution of the downstream tasks, which is the right thing to do, but the latter blocks the submission of the downstream tasks and unrolling of the program's control flow, which should be avoided at all cost. I think you should convert it to an index task not to stall the runtime pipeline. it defeats the purpose of deferred execution.

> As long as I remove the issue_execution_fence, it will not block the Legion runtime to progress tasks on other threads.

no this is not true. what `issue_execution_fence(block=True)` really does is that it blocks on a future returned by the fence. so it's basically the same thing as you calling `wait` on a future from the task. so both the blocking on a fence and the `wait` call on a future block the progress of the program.